flex_model/core/wrapper.py

```diff
@@ -133,46 +181,162 @@ def __init__(
         """
         super().__init__()
         self.module = module
-        self.hook_functions: Dict[str, HookFunction] = {}
-        self._hook_function_handles: Dict[str, torch.utils.hooks.RemovableHandle] = {}
-        self._hooks_active: bool = False
+        self.hook_functions: Dict[
+            nn.Module, Dict[HookFunction, torch.utils.hooks.RemovableHandle]
+        ] = {"all": defaultdict(dict)}
         self.output_ptr = output_ptr
         self.save_ctx: Namespace = Namespace()  # dumb Namespace for now
         self.trainable_modules = nn.ModuleDict()
         self.tp_size = tensor_parallel_size
         self.pp_size = pipeline_parallel_size
         self.dp_size = data_parallel_size
 
+        # Initialize FM distributed.
         if torch.distributed.is_initialized():
             world_size = self.tp_size * self.pp_size * self.dp_size
             dist.initialize_distributed_backend(
                 world_size, self.tp_size, self.pp_size, self.dp_size,
             )
             dist.initialize_activation_parallel()
 
-    def register_hook_function(self, hook_function: HookFunction) -> None:
-        """Register a user-defined `HookFunction` instance.
+        global _OFFLOAD_MODES
+        assert offload_mode in _OFFLOAD_MODES
+        self.offload_mode = offload_mode
 
-        Given a :class:`HookFunction` attach any necessary global context such as
-        the activation output dictionary. Save it into the :class:`FlexModel`s
-        :class:`HookFunction` collection keyed by the module name to hook into.
+        # Create shared state between FM instance and HF instances.
+        self._shared_state = _SharedState(
+            self.output_ptr, self.save_ctx, self.trainable_modules, self.offload_mode,
+        )
 
-        :param HookFunction hook_function: User-defined :class:`HookFunction` instance
-            to register.
-        """
+        # Initialize mappings.
+        self._hook_type_to_pt_attr = {
+            "forward": "register_forward_hook",
+            "full_backward": "register_full_backward_hook",
+            "tensor": "register_hook",
+            "forward_pre": "register_forward_pre_hook",
+            "backward_pre": "register_full_backward_pre_hook",
+        }
```

Comment:
This is also defined in `HookFunction`. There are many mappings defined in both classes, maaybe find a way to unify these for better code reuse